### PR TITLE
Add an implementation for C# composition types

### PR DESCRIPTION
**Pull request by cicciodm**
🚨 **IMPORTANT: Please do not create a Pull Request without creating an issue first.**

*Any change needs to be discussed before proceeding. Failure to do so may result in the rejection of the pull request.*

## Description

Adding an implementation of Composition Types (Union Types and Interfaces) for the c# plugin, as discussed in the related issue.
Related # (issue)
dotansimha/graphql-code-generator-community#149 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Screenshots/Sandbox (if appropriate/relevant):

Adding links to sandbox or providing screenshots can help us understand more about this PR and take action on it as appropriate

## How Has This Been Tested?

Added tests related to composition types in test snapshots

**Test Environment**:
- OS: Win10
- `@graphql-codegen/...`: fresh master
- NodeJS: 14.15

## Checklist:

- [x] I have followed the [CONTRIBUTING](https://github.com/the-guild-org/Stack/blob/master/CONTRIBUTING.md) doc and the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by n1ru4l** [comment]
Hey, there! Can you please rebase this? What is missing before we can merge this?

**Comment by HerrNiklasRaab** [comment]
@cicciodm Any updates on this?

**Comment by cicciodm** [comment]
@HerrNiklasRaab I have been completely swamped with work over the last month, I will try to address comments and provide a new iteration during the weekend.

File: packages/plugins/c-sharp/c-sharp/src/visitor.ts
Comment (cicciodm) [reply packages/plugins/c-sharp/c-sharp/src/visitor.ts:240]: Looking back at this, and I realised that having `Newtonsoft.Json` installed is a pre-requisite for getting this implementation to work. We are using `JToken.ToObject` as an intermediate step for deserializing CompositionType object in their correct concrete implementation. Do you have any suggestion on how this could be changed?

Comment (seclerp) [reply packages/plugins/c-sharp/c-sharp/src/visitor.ts:239]: @cicciodm 
We need to avoid such dependency. Current implementation forces users to use `Newtonsoft.Json` and it's not a good idea (in case if a user doesn't want to use any JSON stuff or use STJ).

So I see 2 options here:
1. Make converters logic as an additional configuration option for C# plugin settings to enable/disable such converters generation (in the case when `Newtonsoft.Json` is used);
2. Move converters logic to a separate NuGet package and then create a configuration option to this plugin to use your NuGet while generating (also Newtonsoft-specific option).

Personally, I think the second approach is better because it removes copy-pasted stuff from type to type and uses 1 centralized NuGet. But it also creates some responsibility to you for supporting such a library.

Comment (cicciodm) [reply packages/plugins/c-sharp/c-sharp/src/visitor.ts:239]: @seclerp I have done some research, and I believe that there is a possibility to re-implement the custom parser only by using `System.Text.Json`, as the library provides an intermediate representation of untyped deserialized object, a generic `Deserialize<T>` method that could be used for deserialization, and custom Deserializer for composition type.

In theory, the dependency on `Newtonsoft.Json` can be removed.

In practice, I would not be able to do the work and testing necessary to get this to work before the holidays, or next January even.

I think that a viable path forward is to perform a check on the `Newtonsoft.Json` dependency and use the current implementation only on a successful check: this could unblock a number of devs in the short term. I can then add another issue for removing this dependency, and get to it in the future. Thoughts?

Comment (seclerp) [reply packages/plugins/c-sharp/c-sharp/src/visitor.ts:239]: @cicciodm in my previous message I was not clear with what I mean by "dependency" in our context. `Newtonsoft.Json` is not a problem in general, because other plugin code already using it for other purposes (basic serialization attributes, for example). The problem is in your usage of Newtonsoft: it is used for any type of configuration, even if a user chose `System.Text.Json` attributes, it will still generate Newtonsoft-related code and it will fall to compile on the user machine (because Newtonsoft NuGet should be installed in such case)

I agree with your proposal about temporal enabling your feature just for the `Newtonsoft.Json` configuration. But could you introduce configuration options for that? See attributes configuration for examples

Comment (cicciodm) [reply packages/plugins/c-sharp/c-sharp/src/visitor.ts:239]: I understand what you mean, rather than saying "The dependency can be removed" I should have phrased it as "I can build an alternative based on `System.Text.Json`" 👍 

Adding a configuration option such as `emitCompositionTypes` and adding documentation about having to install the `Newtonsoft.Json` NuGet makes sense to me!

**Comment by petrkoutnycz** [comment]
This would be really nice to have. Currently C# plugin generates invalid code that cannot be compiled.

**Comment by cicciodm** [comment]
I will have time to continue the work on this PR this week! Thanks for your patience :D

**Comment by cicciodm** [comment]
@seclerp Is there any other issue I should address on this PR?

**Comment by seclerp** [comment]
LGFM

**Review by seclerp — APPROVED** [review]

**Comment by cicciodm** [comment]
Thank you @seclerp! I can see that Vercel had a failure during deployment, do you know who could assist in completing the PR?

**Comment by seclerp** [comment]
@dotansimha could you take a look, please?

**Comment by cicciodm** [comment]
Just pushed some fixes after merging with master, this looks ready to check in now :)

**Comment by JulianKeppelerCP** [comment]
Any updates on this? Would need this fix too

**Comment by cicciodm** [comment]
@dotansimha Would you have the chance to look into this PR?

File: packages/plugins/c-sharp/c-sharp/src/visitor.ts
Comment (azulus) [inline packages/plugins/c-sharp/c-sharp/src/visitor.ts:93]: Is YammerGQLTypes what you want here? I was applying your change against master to create a local patch and it's breaking here.

Comment (jharwood0) [reply packages/plugins/c-sharp/c-sharp/src/visitor.ts:93]: There might be more correct ways of doing this but I just stumbled across this and found this worked for me in a pinch:

```
diff --git a/packages/plugins/c-sharp/c-sharp/src/visitor.ts b/packages/plugins/c-sharp/c-sharp/src/visitor.ts
index ad6af2859..9590d289d 100644
--- a/packages/plugins/c-sharp/c-sharp/src/visitor.ts
+++ b/packages/plugins/c-sharp/c-sharp/src/visitor.ts
@@ -79,7 +79,7 @@ export class CSharpResolversVisitor extends BaseVisitor<CSharpResolversPluginRaw
 private static ConcurrentDictionary<string, Func<JToken, object>> ToObjectForTypenameCache = new ConcurrentDictionary<string, Func<JToken, object>>();
   `;
 
-  private compositionTypeConvertersBlock = `
+  private compositionTypeConvertersBlock = (namespace: string, className: string) => `
 /// <summary>
 /// Given __typeName returns JToken::ToObject[__typeName]. (via cache to improve performance)
 /// </summary>
@@ -90,7 +90,7 @@ public static Func<JToken, object> GetToObjectMethodForTargetType(string typeNam
   if (!ToObjectForTypenameCache.ContainsKey(typeName))
   {
     // Get the type corresponding to the typename
-    Type targetType = typeof(YammerGQLTypes).Assembly
+    Type targetType = typeof(${namespace}.${className}).Assembly
       .GetTypes()
       .ToList()
       .Where(t => t.Name == typeName)
@@ -262,7 +262,7 @@ public class CompositionTypeListConverter : JsonConverter
 
   public addCompositionTypeConverterDefinitions(blocks: string[]): string[] {
     return this._parsedConfig.emitCompositionTypes
-      ? [this.compositionTypeCacheDefinition, ...blocks, this.compositionTypeConvertersBlock]
+      ? [this.compositionTypeCacheDefinition, ...blocks, this.compositionTypeConvertersBlock(this.config.namespaceName, this.config.className)]
       : blocks;
   }
 

```

Comment (cicciodm) [reply packages/plugins/c-sharp/c-sharp/src/visitor.ts:93]: Thanks for spotting this, @jharwood0 patch seems to make sense, I will test it out myself and push another iteration!

Comment (charlypoly) [reply packages/plugins/c-sharp/c-sharp/src/visitor.ts:93]: @cicciodm, I'm staying at your disposal on The Guild side to help you make this PR move forward!

**Comment by joshlrogers** [comment]
@cicciodm anywhere I can send you a coffee, so you might be able to finish this? :)

**Comment by charlypoly** [comment]
Hi!

All community plugins (see list below) have been moved to a new [`dotansimha/graphql-code-generator-community`](https://github.com/dotansimha/graphql-code-generator-community).
<details>
<summary>Community plugins:</summary>

- `@graphql-codegen/typescript-react-apollo`
- `@graphql-codegen/typescript-graphql-request`
- `@graphql-codegen/typescript-apollo-angular`
- `@graphql-codegen/typescript-apollo-client-helpers`
- `@graphql-codegen/typescript-react-query`
- `@graphql-codegen/typescript-urql`
- `@graphql-codegen/named-operations-object`
- `@graphql-codegen/urql-introspection`
- `@graphql-codegen/flow-resolvers`
- `@graphql-codegen/typescript-vue-apollo`
- `@graphql-codegen/typescript-rtk-query`
- `@graphql-codegen/flow-operations`
- `@graphql-codegen/typescript-msw`
- `@graphql-codegen/typescript-mongodb`
- `@graphql-codegen/typescript-type-graphql`
- `@graphql-codegen/jsdoc`
- `@graphql-codegen/typescript-vue-urql`
- `@graphql-codegen/kotlin`
- `@graphql-codegen/typescript-vue-apollo-smart-ops`
- `@graphql-codegen/java`
- `@graphql-codegen/c-sharp-operations`
- `@graphql-codegen/hasura-allow-list`
- `@graphql-codegen/typescript-stencil-apollo`
- `@graphql-codegen/relay-operation-optimizer`
- `@graphql-codegen/typescript-oclif`
- `@graphql-codegen/java-resolvers`
- `@graphql-codegen/java-apollo-android`
- `@graphql-codegen/flutter-freezed`

</details>

### Please move your PR by following these steps:

From your current PR branch:
- git remote add codegen-community git@github.com:dotansimha/graphql-code-generator-community.git
- git fetch codegen-community
- git push codegen-community

Open the PR on [`dotansimha/graphql-code-generator-community`](https://github.com/dotansimha/graphql-code-generator-community) and close this PR.

Thank you!

**Pull request closed without merging**
